src/transformers/tokenization_utils_base.py

```diff
@@ -1877,8 +1877,13 @@ def apply_chat_template(
                 final_message = chat[-1]["content"]
                 if isinstance(final_message, (list, tuple)):
                     final_message = final_message[-1]["text"]
-                final_message = final_message.strip()
-                rendered_chat = rendered_chat[: rendered_chat.rindex(final_message) + len(final_message)].rstrip()
+                try:
+                    final_message = final_message
+                    rendered_chat = rendered_chat[: rendered_chat.rindex(final_message) + len(final_message)]
+                except:  # noqa: E722
+                    # Some chat templates like Llama-3.1 trim messages before rendering, so we must do the same here.
+                    final_message = final_message.strip()
+                    rendered_chat = rendered_chat[: rendered_chat.rindex(final_message) + len(final_message)].rstrip()
```

Comment:
FYI @Rocketknight1 I kept the `rstrip()` here for consistency with the prior implementation. Let me know if you want it removed.

Reply:
It should make no difference, because we're slicing the string to the end of `final_message.strip()`, which must end with a non-whitespace character because it's been stripped. Therefore, there'll be nothing to `rstrip()` regardless! As such, we can remove it for cleanliness, but it won't break anything if you leave it either.

Reply:
ok let's nuke it!